### PR TITLE
Maintenance: update FadingCounter API

### PR DESCRIPTION
**Pull request by yadij**

### CODE REVIEW

File: src/FadingCounter.cc [hunk 1]
```diff
@@ -14,51 +14,53 @@
...
         counters.push_back(0);
 }
 
-void FadingCounter::clear()
+void
+FadingCounter::clear()
 {
-    for (int i = 0; i < precision; ++i)
+    for (size_t i = 0; i < Precision; ++i)
         counters[i] = 0;
```
Comment (kinkie) [inline src/FadingCounter.cc:28]: how about
```suggestion
    std::fill(counters.begin(), counters.end(), 0);
```
?

File: src/FadingCounter.cc [hunk 2]
```diff
@@ -14,51 +14,53 @@
...
 
     const double deltas = (current_dtime - lastTime) / delta;
-    if (deltas >= precision || current_dtime < lastTime) {
+    if (deltas >= Precision || current_dtime < lastTime) {
         clear(); // forget all values
     } else {
         // forget stale values, if any
         // fmod() or "current_dtime/delta" will overflow int for small deltas
-        const int lastSlot = static_cast<int>(fmod(lastTime, horizon) / delta);
+        const int lastSlot = static_cast<int>(fmod(lastTime, horizon()) / delta);
```
Comment (kinkie) [inline src/FadingCounter.cc:60]:
```suggestion
        const auto lastSlot = static_cast<int>(fmod(lastTime, horizon()) / delta);
```

File: src/FadingCounter.cc [hunk 3]
```diff
@@ -14,51 +14,53 @@
...
         const int staleSlots = static_cast<int>(deltas);
         for (int i = 0, s = lastSlot + 1; i < staleSlots; ++i, ++s) {
-            const int idx = s % precision;
+            const int idx = s % Precision;
```
Comment (kinkie) [inline src/FadingCounter.cc:63]:
```suggestion
            const auto idx = s % Precision;
```

File: src/FadingCounter.cc [hunk 4]
```diff
@@ -67,8 +69,8 @@ int FadingCounter::count(int howMany)
 
     // apply new information
     lastTime = current_dtime;
-    const int curSlot = static_cast<int>(fmod(lastTime, horizon) / delta);
-    counters[curSlot % precision] += howMany;
+    const int curSlot = static_cast<int>(fmod(lastTime, horizon()) / delta);
```
Comment (kinkie) [inline src/FadingCounter.cc:72]:
```suggestion
    const auto curSlot = static_cast<int>(fmod(lastTime, horizon()) / delta);
```

File: src/FadingCounter.cc [hunk 5]
```diff
@@ -14,51 +14,53 @@
 
 #include <cmath>
 
-FadingCounter::FadingCounter(): horizon(-1), precision(10), delta(-1),
-    lastTime(0), total(0)
+FadingCounter::FadingCounter()
```
Comment (kinkie) [inline src/FadingCounter.cc:17]:
```suggestion
FadingCounter::FadingCounter() : counters(Precision,0)
```

File: src/FadingCounter.cc [hunk 6]
```diff
@@ -14,51 +14,53 @@
...
 {
-    counters.reserve(precision);
-    while (counters.size() < static_cast<unsigned int>(precision))
+    counters.reserve(Precision);
+    while (counters.size() < Precision)
         counters.push_back(0);
```
Comment (kinkie) [inline src/FadingCounter.cc:21]: The initializer in the previous line would make this moot

```suggestion
```

**Review by rousskov — CHANGES_REQUESTED** [review]
AFAICT, this PR introduces runtime bugs, but I am focusing this review on primary API changes: If my horizon type recommendations are followed, those bugs may go away automagically.

Please test FadingCounter configurations with `-1`, `0`, and `1 second` horizon parameters before requesting additional reviews.

File: src/FadingCounter.h [hunk 7]
```diff
@@ -15,26 +15,28 @@
 class FadingCounter
 {
 public:
+    static const size_t Precision = 10; ///< #counting slots, controls measur. accuracy
```
[the remark this answers is not in the document]
Comment (yadij) [reply src/FadingCounter.h:18]: Your comment implies that this is something new. It is not `precision` has always existed and never modified after object initialization to a value of `10`. Making it a `static const` just reduces the needless memory consumption.

File: src/FadingCounter.h [hunk 8]
```diff
@@ -15,26 +15,28 @@
...
+
     FadingCounter();
 
     /// 0=remember nothing; -1=forget nothing; new value triggers clear()
-    void configure(double horizonSeconds);
+    void configure(const time_t horizonSeconds);
```
[the remark this answers is not in the document]
Comment (yadij) [reply src/FadingCounter.h:23]: One step at a time. Using a `std::chrono` here requires `squid.conf` parser logic and modifications to the internal logic of `FadingCounter`. They are explicitly being left to followup PR.

File: src/FadingCounter.h [hunk 9]
```diff
@@ -15,26 +15,28 @@
...
 
     void clear(); ///< forgets all events
 
     int count(int howMany); ///< count fresh, return #events remembered
```
[the remark this answers is not in the document]
Comment (yadij) [reply src/FadingCounter.h:27]: If you insist, I was going to leave it to a followup as there is more impact on callers with a change to `count()`.

**Comment by yadij** [comment]
> Please test FadingCounter configurations with `-1`, `0`, and `1 second` horizon parameters before requesting additional reviews.

Please add the missing unit-tests for `FadingCounter`. I have ensured these changes are bug-compatible with existing code.

File: src/FadingCounter.h (hunk 7 above)
[the remark this answers is not in the document]
Comment (yadij) [reply src/FadingCounter.h]: no reason. reverted now

**Comment by rousskov** [comment]
> > Please test FadingCounter configurations with `-1`, `0`, and `1 second` horizon parameters before requesting additional reviews.
> 
> Please add the missing unit-tests for `FadingCounter`. I have ensured these changes are bug-compatible with existing code.

Personally, I do not plan on adding unit-tests for FadingCounter (and I am not sure I agree that they are "missing").

> I have ensured these changes are bug-compatible with existing code.

This request is not about existing official code. I strongly suspect that the current PR introduces _new_ bugs in FadingCounter implementation. If my suspicions are correct, the requested tests should expose those bugs (and prevent them from re-occurring as we refactor PR code to address change requests). I am not asking for anything complex or permanent like unit tests. I do not even need to see your test code.

File: src/FadingCounter.h (hunk 7 above)
[the remark this answers is not in the document]
Comment (rousskov) [reply src/FadingCounter.h]: > Your comment implies that this is something new. It is not `precision` has always existed and never modified after object initialization to a value of `10`.

That "something new" here is a design change from "each counter may have its own precision (that never changes)" design to "all counters have identical precision (that never changes)" design. The former design does not allow std::array usage. The new design does. The fact that the old design used a constant 10 to initialize that per-counter precision is not important (from the design point of view).

BTW, each design has pros/cons, of course. I am not attacking or defending a particular design in this change request. I am only saying that if this PR switches to a one-size-fits-all design, it should go much further.

> Making it a `static const` just reduces the needless memory consumption.

I disagree with the "just" part, and the next steps I outlined would remove a lot more waste than an integer data member! Also, such changes to private members seem to be outside this "update API" PR scope. This PR does not have to make this change, but if it does, it becomes responsible for seeing it through.

File: src/FadingCounter.h (hunk 8 above)
[the remark this answers is not in the document]
Comment (rousskov) [reply src/FadingCounter.h:21]: > Using a `std::chrono` here requires `squid.conf` parser logic 

I seriously doubt that using std::chrono for FadingCounter::horizon requires any squid.conf parser changes, but, if it really does, then FadingCounter::horizon type and similar changes are not compatible with an "Update FadingCounter API" PR (because they require too many out-of-scope changes).

> and modifications to the internal logic of `FadingCounter`.

I do not think so, but our definitions of "internal logic" may differ.

> They are explicitly being left to followup PR.

In summary, it is totally fine to postpone conversion of FadingCounter internals to std::chrono beyond this PR as long as this PR does not change FadingCounter::horizon type (or contains other similar internal FadingCounter changes that are difficult to do correctly, especially without std::chrono help).

File: src/FadingCounter.h (hunk 9 above)
[the remark this answers is not in the document]
Comment (rousskov) [reply src/FadingCounter.h:25]: If the scope of this PR is to spruce up this old class interface, then I insist. If the scope is different but remains unknown to me, then I cannot tell whether this PR has to include these changes until the scope is clarified. Either way, the ball is on your side.

[the remark this answers is not in the document]
Comment (yadij) [reply src/FadingCounter.h]: Done in [d9ad336](https://github.com/squid-cache/squid/pull/1782/commits/d9ad3367ee4c7cc4c7bd5c29524f2946e3b66cec)

File: src/FadingCounter.h (hunk 8 above)
[the remark this answers is not in the document]
Comment (yadij) [reply src/FadingCounter.h:21]: `FadingCounter::horizon` is set by several configuration options. Currently they use `parse_time()`. I do not want to extend scope of this specific PR outside of `FadingCounter`.

File: src/FadingCounter.cc [hunk 10]
```diff
@@ -14,63 +14,53 @@
...
-    counters.reserve(precision);
-    while (counters.size() < static_cast<unsigned int>(precision))
-        counters.push_back(0);
-}
-
-void FadingCounter::clear()
-{
-    for (int i = 0; i < precision; ++i)
-        counters[i] = 0;
+    std::fill(counters.begin(), counters.end(), 0);
```
Comment (rousskov) [inline src/FadingCounter.cc:20]: Please simplify this code, similar to code this PR already uses to initialize `counters` during object construction:

```suggestion
    counters = {};
```

File: src/FadingCounter.cc [hunk 11]
```diff
@@ -14,63 +14,53 @@
...
         // fmod() or "current_dtime/delta" will overflow int for small deltas
-        const int lastSlot = static_cast<int>(fmod(lastTime, horizon) / delta);
+        const auto lastSlot = static_cast<int>(fmod(lastTime, horizon()) / delta);
         const int staleSlots = static_cast<int>(deltas);
         for (int i = 0, s = lastSlot + 1; i < staleSlots; ++i, ++s) {
-            const int idx = s % precision;
+            const auto idx = s % counters.size();
             total -= counters[idx];
             counters[idx] = 0;
-            Must(total >= 0);
```
Comment (rousskov) [inline src/FadingCounter.cc:64]: This assertion can be preserved if it is moved before `total` is decremented and adjusted to look like `Assure(total >= counters[idx])`.

**Review by rousskov — CHANGES_REQUESTED** [review]

File: src/FadingCounter.h (hunk 7 above)
[the remark this answers is not in the document]
Comment (rousskov) [reply src/FadingCounter.h]: > > "each counter may have its own precision (that never changes)"
> 
> There is/was no setter for the `private` member. So no, they could not have changed it from the constructed value of `10`. It has always been a constant but lacked the `const` marker.

The design aspect discussed here is not about _changing_ existing precision of a FadingCounter instance. It is about configuring custom precision at FadingCounter construction time. With the old design, it would be trivial to add a constructor (parameter) that initializes precision to a value other than 10 for a particular FadingCounter instance. The precision member would still be constant for that instance -- it will never change. With the new design, that customization becomes impossible because the precision value is shared across all FadingCounter instances.

> Your line of arguments are based on a design which several years worth of development have yet to identify a single case of need for the `precision` member to be non-const.

The above sounds as if I am defending the old design and/or trying to preserve it. I am not! My argument is not "the old design was good, so this PR should ...". Instead, my argument is "this PR _changes_ the old design, so this PR is responsible for the consequences of that change". Migrating from runtime-configurable std::vector to compiler-sized std::array is such a consequence.

Thank you for addressing my concern.

File: src/FadingCounter.cc (hunk 11 above)
Comment (yadij) [reply src/FadingCounter.cc:64]: I removed this and the one below because they seem to exist purely for paranoia reasons and do not appear to be necessary any longer.

File: src/FadingCounter.cc (hunk 10 above)
Comment (yadij) [reply src/FadingCounter.cc]: Done in [c60bf4c](https://github.com/squid-cache/squid/pull/1782/commits/c60bf4cce828a7f5ac37876def190de9db4fa9e1)

File: src/FadingCounter.h (hunk 8 above)
[the remark this answers is not in the document]
Comment (yadij) [reply src/FadingCounter.h]: Fine. Done in [c60bf4c](https://github.com/squid-cache/squid/pull/1782/commits/c60bf4cce828a7f5ac37876def190de9db4fa9e1).

File: src/FadingCounter.h [hunk 12]
```diff
@@ -9,32 +9,41 @@
...
-    /// read-only memory horizon in seconds; older events are forgotten
-    double horizon;
+    /// read-only memory horizon; older events are forgotten
+    Horizon horizon() const { return horizon_; }
 
 private:
-    const int precision; ///< #counting slots, controls measur. accuracy
-    double delta; ///< sub-interval duration = horizon/precision
+    Horizon horizon_ = Horizon::max();
+    double delta = -1; ///< sub-interval duration = horizon/precision
```
Comment (rousskov) [inline src/FadingCounter.h:42]: Please use std::chrono::duration-based type for this horizon-derived cached value. While you adjust this line, please update the data member description to avoid using the now-gone `precision` field.

**Review by rousskov — CHANGES_REQUESTED** [review]
> Alex: AFAICT, this PR introduces runtime bugs, but I am focusing this review on primary API changes: If my horizon type recommendations are followed, those bugs may go away automagically.

Those old concerns remain. It is still possible that they will disappear if horizon-related  adjustments go far enough; I have added another specific change request in hope to make further progress in this direction.

> Alex: Please test FadingCounter configurations with -1, 0, and 1 second horizon parameters before requesting additional reviews.

I doubt this was done because PR code [does not even build](https://github.com/squid-cache/squid/actions/runs/14357088373/job/40248962387?pr=1782).
